pypesto/ensemble/ensemble.py

```diff
@@ -367,6 +371,155 @@ def from_sample(
+        return Ensemble(x_vectors=x_vectors,
+                        x_names=result.problem.x_names,
+                        vector_tags=vector_tags,
+                        lower_bound=result.problem.lb_full,
+                        upper_bound=result.problem.ub_full,
+                        **kwargs)
+
+    @staticmethod
+    def from_history(
+            result: Result,
+            cutoff: float = np.inf,
+            max_size: int = np.inf,
+            max_per_start: int = np.inf,
+            **kwargs,
+    ):
+        """Construct an ensemble from the history of an optimization.
+
+        Parameters
+        ----------
+        result:
+            A pyPESTO result that contains an optimization result
+            with history recorded.
+        cutoff:
+            Exclude parameters from the optimization if the nllh
+            is higher than the `cutoff`.
+        max_size:
+            The maximum size the ensemble should be.
+        max_per_start:
+            The maximum number of vectors to be included from a
+            single optimization start.
+
+        Returns
+        -------
+        The ensemble.
+        """
+        if not result.optimize_result.list[0].history.options['trace_record']:
+            logger.warning('The optimize result has no trace. The Ensemble '
+                           'will automatically be created through '
+                           'from_optimization().')
+            return Ensemble.from_optimization_endpoints(result=result,
+                                                        cutoff=cutoff,
+                                                        max_size=max_size,
+                                                        **kwargs)
+        x_vectors = []
+        vector_tags = []
+        x_names = result.problem.x_names
+        lb = result.problem.lb_full
+        ub = result.problem.ub_full
+
+        # calculate the number of starts whose final nllh is below cutoff
+        n_starts = 0
+        for start in result.optimize_result.list:
+            if start['fval'] > cutoff:
+                break
+            n_starts += 1
+
+        if n_starts*max_per_start > max_size:
+            logger.info(f'The number of starts that can contribute an '
+                        'ensemble vector multiplied with max_per_start '
+                        'is higher than max_size. Thus we will lower '
+                        'max_per_start. If you do not want this to '
+                        'happen consider increasing max_size to '
+                        f'{max_per_start*n_starts} or decrease cutoff.')
+            max_per_start = math.floor(max_size/n_starts)
+
+        for i in range(n_starts):
+            trace_x = \
+                result.optimize_result.list[i]['history'].get_x_trace()
+            trace_fval = \
+                result.optimize_result.list[i]['history'].get_fval_trace()
+
+            # calculate number of candidates
+            n_cand = 0
```

Comment:
I see the nice "fail fast" aspect of this implementation, but personally, I'd go for some list comprehension here...
`n_cand = sum([iter_fval for iter_fval in trace_fval if iter_fval < cutoff])`. That saves four lines...

Reply:
if I sum that up, wouldn't that be then the sum of the fvals? I think `n_can = len([iter_fval for iter_fval in trace_fval if iter_fval < cutoff])` would be correct here then?

Reply:
You're right. It should be either something using len or summing over 1!

Reply:
the issue here is that it will sum over all possible iterations, we have that the iterations might not be seamless, as the optimization is not strictly decreasing.

Reply:
```python
>>> import numpy as np
>>> b = np.array([1,2,3,4,5,6,4,3,2,1])
>>> np.sum(b < 5)
8
>>> np.where(np.diff(b < 5))[0][0] + 1
4
```

So setting `n_cand` to the last expression is probably what you want? With some comment to describe what it does since it might be confusing to someone unfamiliar with the code

Reply:
instead of funny where[0][0], consider also numpy nonzero and numpy flatnonzero ;)

Reply:
Also `np.diff(b<5).argmax()`